src/main/core/support/configuration.rs

```diff
@@ -707,13 +707,58 @@ impl std::ops::Deref for Quantity {
     }
 }
 
-#[derive(Debug, Clone, Serialize, Deserialize, JsonSchema)]
+#[derive(Debug, Clone, Serialize, JsonSchema)]
 #[serde(untagged)]
 pub enum ProcessArgs {
     List(Vec<String>),
     Str(String),
 }
 
+/// Serde doesn't provide good deserialization error messages for untagged enums, so we implement
+/// our own. For example, if serde finds a yaml value such as 4 for the process arguments, it won't
+/// deserialize it to the string "4" and the yaml parsing will fail. The serde-generated error
+/// message will say something like "data did not match any variant of untagged enum ProcessArgs at
+/// line X column Y" which isn't very helpful to the user, so here we try to give a better error
+/// message.
+impl<'de> serde::Deserialize<'de> for ProcessArgs {
+    fn deserialize<D: serde::Deserializer<'de>>(deserializer: D) -> Result<Self, D::Error> {
+        let value = serde_yaml::Value::deserialize(deserializer)?;
+
+        if let Ok(list) = Vec::<String>::deserialize(value.clone()) {
+            return Ok(Self::List(list));
+        }
+
+        if let Ok(s) = String::deserialize(value.clone()) {
+            return Ok(Self::Str(s));
+        }
+
+        let unexpected = match value {
+            serde_yaml::Value::Null => serde::de::Unexpected::Unit,
+            serde_yaml::Value::Bool(b) => serde::de::Unexpected::Bool(b),
+            serde_yaml::Value::Number(n) if n.is_u64() => {
+                serde::de::Unexpected::Unsigned(n.as_u64().unwrap())
+            }
+            serde_yaml::Value::Number(n) if n.is_i64() => {
+                serde::de::Unexpected::Signed(n.as_i64().unwrap())
+            }
+            serde_yaml::Value::Number(n) if n.is_f64() => {
+                serde::de::Unexpected::Float(n.as_f64().unwrap())
+            }
+            serde_yaml::Value::Number(_) => {
+                unimplemented!("Did serde_yaml add a new number type?")
+            }
+            serde_yaml::Value::String(ref s) => serde::de::Unexpected::Str(s),
+            serde_yaml::Value::Sequence(_) => serde::de::Unexpected::Seq,
+            serde_yaml::Value::Mapping(_) => serde::de::Unexpected::Map,
```

Comment:
I *think* this looks fine, but I'm confused how we get here for anything other than a `Mapping`; any other valid yaml value should be able to be parsed as a `String` or `Vec<String>`, shouldn't it? e.g. in your example, shouldn't the value `4` be successfully parsed as a string "4"?  Or do bare integers need to be quoted to be considered a valid string?

Reply:
With Serde you get one chance to deserialize something. If you try to deserialize to a `Vec<String>` and it's unsuccessful, you can't retry again with a `String`. To get this functionality, you need to first deserialize to an intermediate type like [`serde_yaml::Value`](https://docs.serde.rs/serde_yaml/enum.Value.html). Since this type implements `Clone`, you can keep cloning it and trying to deserialize it to different types. But once you've deserialized to `serde_yaml::Value`, you've lost the original representation. For example if you deserialize to a `serde_yaml::Value::Bool`, there's no 1:1 correlation with a `String`. Should `serde_yaml::Value::Bool(true)` be `"true"`, `"TRUE"`, or `"True"` (which are all valid yaml booleans)?